src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java

```diff
@@ -34,176 +35,131 @@
-        assertEquals(TIMEOUT, e.getTimeout());
-        assertEquals(TimeUnit.MILLISECONDS, e.getTimeUnit());
-    }
-
-    private ThrowingRunnable evaluateWithDelegate(final Statement delegate) {
-        return new ThrowingRunnable() {
-            public void run() throws Throwable {
-                statement.nextStatement = delegate;
-                statement.waitDuration = 0;
-                failOnTimeout.evaluate();
-            }
-        };
-    }
-
-    private ThrowingRunnable evaluateWithException(Exception exception) {
-        return evaluateWithDelegate(new Fail(exception));
-    }
-
-    private ThrowingRunnable evaluateWithWaitDuration(final long waitDuration) {
-        return new ThrowingRunnable() {
-            public void run() throws Throwable {
-                statement.nextStatement = null;
-                statement.waitDuration = waitDuration;
-                failOnTimeout.evaluate();
-            }
-        };
-    }
-
-    private static final class TestStatement extends Statement {
-        long waitDuration;
-
-        Statement nextStatement;
-
-        @Override
-        public void evaluate() throws Throwable {
-            sleep(waitDuration);
-            if (nextStatement != null) {
-                nextStatement.evaluate();
-            }
-        }
+                run(failAfter50Ms(new RunForASecond())));
+        assertEquals(50, e.getTimeout());
+        assertEquals(MILLISECONDS, e.getTimeUnit());
     }
 
     @Test
-    public void stopEndlessStatement() throws Throwable {
-        InfiniteLoopStatement infiniteLoop = new InfiniteLoopStatement();
-        FailOnTimeout infiniteLoopTimeout = builder().withTimeout(TIMEOUT, MILLISECONDS).build(infiniteLoop);
-        try {
-            infiniteLoopTimeout.evaluate();
-        } catch (Exception timeoutException) {
-            sleep(20); // time to interrupt the thread
-            int firstCount = InfiniteLoopStatement.COUNT;
-            sleep(20); // time to increment the count
-            assertTrue("Thread has not been stopped.",
-                    firstCount == InfiniteLoopStatement.COUNT);
-        }
-    }
-
-    private static final class InfiniteLoopStatement extends Statement {
-        private static int COUNT = 0;
-
-        @Override
-        public void evaluate() throws Throwable {
-            while (true) {
-                sleep(10); // sleep in order to enable interrupting thread
-                ++COUNT;
-            }
-        }
+    public void statementThatCanBeInterruptedIsStoppedAfterTimeout() throws Throwable {
+        // RunForASecond can be interrupted because it uses Thread.sleep which
+        // can be interrupted.
+        RunForASecond runForASecond = new RunForASecond();
+        assertThrows(
+                TestTimedOutException.class,
+                run(failAfter50Ms(runForASecond)));
+
+        sleep(20); // time to interrupt the thread
+        runForASecond.stillExecuting.set(false);
```

Comment:
I found this to be confusing. The statement may or may not still be executing. It's not clear here that `RunForASecond` has code that will set it to be true if it is still running after this point. 

I suggest adding a `finally` block in the `RunForASecond` that sets `stillExecuting` to `false`.

Alternatively, replace `stillExecuting` with an `AtomicInteger` and rewrite this to be similar to what it was before.

Reply:
I changed the test to use a CountDownLatch: 01dfbc81d49b1c561682119e3eea2c6218b59b44.

Reply:
Good idea. Perhaps there should still be a short sleep in the `while` loop in `RunForASecond`?

Reply:
It works in the CI and locally. Therefore I would like to go without the sleep until we know about a case where it is necessary.